File: src/endstone_endweave/protocol/v924_to_v898/protocol.py
"""Protocol factory for v924 (1.26.0) server <- v898 (1.21.130) client."""

from endstone_endweave.protocol import Protocol
from endstone_endweave.protocol.mappings.v898_v924 import MAPPINGS
from endstone_endweave.protocol.packet_ids import PacketId
from endstone_endweave.protocol.v924_to_v898.handlers.biome_definition_list import (
    rewrite_biome_definition_list,
)
from endstone_endweave.protocol.v924_to_v898.handlers.book_edit import (
    rewrite_book_edit,
)
from endstone_endweave.protocol.v924_to_v898.handlers.camera import (
    rewrite_camera_instruction,
)
from endstone_endweave.protocol.v924_to_v898.handlers.camera_aim_assist import (
    rewrite_camera_aim_assist_presets,
)
from endstone_endweave.protocol.v924_to_v898.handlers.data_store import (
    rewrite_clientbound_data_store,
    rewrite_serverbound_data_store,
)
from endstone_endweave.protocol.v924_to_v898.handlers.debug_drawer import (
    rewrite_debug_drawer,
)
from endstone_endweave.protocol.v924_to_v898.handlers.diagnostics import (
    rewrite_diagnostics,
)
from endstone_endweave.protocol.v924_to_v898.handlers.graphics_parameter_override import (
    rewrite_graphics_parameter_override,
)
from endstone_endweave.protocol.v924_to_v898.handlers.start_game import (
    rewrite_start_game,
)
from endstone_endweave.protocol.v924_to_v898.handlers.text import (
    rewrite_text_clientbound,
    rewrite_text_serverbound,
)
from endstone_endweave.rewriter import SoundRewriter

SERVER_PROTOCOL = 924
CLIENT_PROTOCOL = 898

def create_protocol() -> Protocol:
    """Create a protocol for v924 server <- v898 client."""
    protocol = Protocol(server_protocol=SERVER_PROTOCOL, client_protocol=CLIENT_PROTOCOL)

    # Serverbound rewriters
    protocol.register_serverbound(PacketId.TEXT, rewrite_text_serverbound)
    protocol.register_serverbound(PacketId.SERVERBOUND_DATA_STORE, rewrite_serverbound_data_store)
    protocol.register_serverbound(PacketId.BOOK_EDIT, rewrite_book_edit)
    protocol.register_serverbound(PacketId.SERVERBOUND_DIAGNOSTICS, rewrite_diagnostics)

    # Clientbound rewriters -- v924/v898 format differences
    protocol.register_clientbound(PacketId.START_GAME, rewrite_start_game)
    protocol.register_clientbound(PacketId.TEXT, rewrite_text_clientbound)
    protocol.register_clientbound(PacketId.CLIENTBOUND_DATA_STORE, rewrite_clientbound_data_store)
    protocol.register_clientbound(PacketId.CAMERA_AIM_ASSIST_PRESETS, rewrite_camera_aim_assist_presets)
    protocol.register_clientbound(PacketId.BIOME_DEFINITION_LIST, rewrite_biome_definition_list)

    # Clientbound rewriters -- LevelSoundEvent remapping
    sound = SoundRewriter(
        sound_remap=MAPPINGS.sound.cap,
        actor_data_int_remappers={MAPPINGS.actor_data_sound_key: MAPPINGS.sound.cap},
        dropped_actor_data_keys=MAPPINGS.dropped_actor_data_keys,
    )
    sound.register(protocol)

    # Clientbound rewriters -- other v924/v898 changes
    protocol.register_clientbound(PacketId.GRAPHICS_PARAMETER_OVERRIDE, rewrite_graphics_parameter_override)
    protocol.register_clientbound(PacketId.CAMERA_INSTRUCTION, rewrite_camera_instruction)
    protocol.register_clientbound(PacketId.SERVER_SCRIPT_DEBUG_DRAWER, rewrite_debug_drawer)

    # Cancel packets unknown to v898
    protocol.cancel_clientbound(
        PacketId.CLIENTBOUND_DATA_DRIVEN_UI_SHOW_SCREEN,
        PacketId.CLIENTBOUND_DATA_DRIVEN_UI_CLOSE_ALL_SCREENS,
        PacketId.CLIENTBOUND_DATA_DRIVEN_UI_RELOAD,
        PacketId.CLIENTBOUND_TEXTURE_SHIFT,
        PacketId.VOXEL_SHAPES,
        PacketId.CAMERA_SPLINE,
        PacketId.CAMERA_AIM_ASSIST_ACTOR_PRIORITY,
    )

    return protocol

File: src/protocol/v924_to_v898/protocol.cpp
/// v924 -> v898 protocol factory.

#include "endweave/protocol/v924_to_v898/protocol.h"

#include "endweave/codec/types/enums.h"
#include "endweave/protocol/packet_ids.h"

namespace endweave::v924_to_v898 {

namespace {

int remap_sound(int v) {
    // Remap LevelSoundEvent from v924 -> v898 (collapse v924 sounds)
    constexpr int V924_START = 597; // sound_boundaries::v944_start (v924 undefined)
    constexpr int V898_START = 578; // sound_boundaries::v898_start
    if (v >= V924_START)
        return v - 19; // V924_START(597) - V898_START(578) = 19
    if (v >= V898_START)
        return static_cast<int>(V898_START); // clamp to Undefined_v898
    return v;
}

} // namespace

ProtocolState create_protocol() {
    auto id = [](PacketId p) { return static_cast<int>(p); };

    auto p = std::make_unique<Protocol>(SERVER_PROTOCOL, CLIENT_PROTOCOL, "v924_to_v898");

    // TODO: implement handlers
    auto stub = [](PacketWrapper& w) { w.passthrough_all(); };

    // Cancel v924-specific packets unknown to v898
    p->cancel_clientbound({
        id(PacketId::CONTAINER_REGISTRY_CLEANUP),                  // 317
        id(PacketId::MOVEMENT_EFFECT),                             // 318
        id(PacketId::CLIENT_CAMERA_AIM_ASSIST),                    // 321
        id(PacketId::CLIENT_MOVEMENT_PREDICTION_SYNC),             // 322
        id(PacketId::UPDATE_CLIENT_OPTIONS),                       // 323
        id(PacketId::PLAYER_VIDEO_CAPTURE),                        // 324
        id(PacketId::PLAYER_UPDATE_ENTITY_OVERRIDES),              // 325
        id(PacketId::PLAYER_LOCATION),                             // 326
        id(PacketId::CLIENTBOUND_CONTROL_SCHEME_SET),              // 327
        id(PacketId::SERVERBOUND_PACK_SETTING_CHANGE),             // 329
        id(PacketId::CLIENTBOUND_DATA_DRIVEN_UI_SHOW_SCREEN),     // 333
        id(PacketId::CLIENTBOUND_DATA_DRIVEN_UI_CLOSE_ALL_SCREENS), // 334
        id(PacketId::CLIENTBOUND_DATA_DRIVEN_UI_RELOAD),           // 335
        id(PacketId::CLIENTBOUND_TEXTURE_SHIFT),                   // 336
        id(PacketId::VOXEL_SHAPES),                                // 337
        id(PacketId::CAMERA_SPLINE),                               // 338
        id(PacketId::CAMERA_AIM_ASSIST_ACTOR_PRIORITY),            // 339
    });

    // Clientbound handlers (reverse of v898_to_v924 serverbound)
    p->register_clientbound(id(PacketId::TEXT), stub);                    // 9
    p->register_clientbound(id(PacketId::BOOK_EDIT), stub);              // 97
    p->register_clientbound(id(PacketId::SERVERBOUND_DIAGNOSTICS), stub); // 315

    // Serverbound handlers (reverse of v898_to_v924 clientbound)
    p->register_serverbound(id(PacketId::START_GAME), stub);                    // 11
    p->register_serverbound(id(PacketId::TEXT), stub);                           // 9
    p->register_serverbound(id(PacketId::CLIENTBOUND_DATA_STORE), stub);        // 330
    p->register_serverbound(id(PacketId::CAMERA_AIM_ASSIST_PRESETS), stub);     // 320
    p->register_serverbound(id(PacketId::GRAPHICS_PARAMETER_OVERRIDE), stub);  // 331
    p->register_serverbound(id(PacketId::CAMERA_INSTRUCTION), stub);           // 300
    p->register_serverbound(id(PacketId::BIOME_DEFINITION_LIST), stub);        // 122
    p->register_serverbound(id(PacketId::SERVER_SCRIPT_DEBUG_DRAWER), stub);   // 328

    // SoundRewriter
    SoundRewriter sound(remap_sound,
        {{static_cast<int>(ActorDataIDs::HEARTBEAT_SOUND_EVENT), remap_sound}},
        {
            static_cast<int>(ActorDataIDs::AIM_ASSIST_PRIORITY_PRESET_ID),   // 136
            static_cast<int>(ActorDataIDs::AIM_ASSIST_PRIORITY_CATEGORY_ID), // 137
            static_cast<int>(ActorDataIDs::AIM_ASSIST_PRIORITY_ACTOR_ID),    // 138
        });
    sound.register_on(*p);

    return {std::move(p), std::move(sound)};
}

} // namespace endweave::v924_to_v898
